oak-run/src/main/java/org/apache/jackrabbit/oak/run/Downloader.java

```diff
@@ -147,14 +163,49 @@ public ItemResponse call() throws Exception {
             sourceUrl.setConnectTimeout(Downloader.this.connectTimeoutMs);
             sourceUrl.setReadTimeout(Downloader.this.readTimeoutMs);
 
+            // Updating a MessageDigest from multiple threads is not thread safe, so we cannot reuse a single instance.
+            // Creating a new instance is a lightweight operation, no need to increase complexity by creating a pool.
+            MessageDigest md = null;
+            if (Downloader.this.checksumAlgorithm != null && item.checksum != null) {
+                md = MessageDigest.getInstance(Downloader.this.checksumAlgorithm);
+            }
+
             Path destinationPath = Paths.get(item.destination);
             Files.createDirectories(destinationPath.getParent());
-            long size;
+            long size = 0;
             try (ReadableByteChannel byteChannel = Channels.newChannel(sourceUrl.getInputStream());
                  FileOutputStream outputStream = new FileOutputStream(destinationPath.toFile())) {
-                size = outputStream.getChannel()
-                        .transferFrom(byteChannel, 0, Long.MAX_VALUE);
+                ByteBuffer buffer = ByteBuffer.allocate(1024);
+                int bytesRead;
+                while ((bytesRead = byteChannel.read(buffer)) != -1) {
+                    buffer.flip();
+                    if (md != null) {
+                        md.update(buffer);
+                    }
+                    size += bytesRead;
+                    outputStream.getChannel().write(buffer);
+                    buffer.clear();
+                }
```

Comment:
> I'm wondering if the performance of doing the transfer with buffers in the Java heap will not be significantly worse than using `transferFrom`, which in principle can use low-level mechanisms of the JVM/OS.

Low-level copy mechanisms are possible only for copies between local channels. When the source is remote, a similar buffer copy is used:
https://github.com/frohoff/jdk8u-jdk/blob/master/src/share/classes/sun/nio/ch/FileChannelImpl.java#L547-L581

> The buffer of 1024 seems small, could hurt performance. I guess the buffer here should be large enough to read a large chunk of the TPC receive buffer. I'm not sure what is the default size, but it is for sure larger than 1024 bytes. I found this which suggests that the buffer is close to 80KB. Maybe here it is better to increase to 16KB or even 32KB, to reduce the CPU overhead.

It was small indeed. I have executed several runs (see results in the main description) and I see some small improvements when the buffer size is larger. The optimal buffer size for downloading and checksumming a file depends on a variety of factors, such as the size of the file being downloaded, the available memory of the system, the network bandwidth, and other system-specific factors.

The `buffer-size` is now configurable. The default is 8k which should be a safe default buffer size to use. It can be changed if needed.

Reply:
Could you explain better what the results mean:

1.52-SNAPSHOT (8kb) | 58.71267322 | 357
-- | -- | --
1.52-SNAPSHOT (8kb, SHA-256) | 54.7603119 | 383

The version without SHA-256, does it do any checksumming? And if it does not do any checksumming, why is it slower? 

And 1.46 is the current version that uses `transferTo`?

Reply:
The version without SHA-256 does not perform any checksumming. Surprisingly is consistently slower than the one performing the checksum validation.

1.46 is the current version that uses `transferTo`. Please notice it was invoked with:

```java
outputStream.getChannel()
                        .transferFrom(byteChannel, 0, Long.MAX_VALUE);
```

where `Long.MAX_VALUE` is the buffer size.

Reply:
>The buffer-size is now configurable. The default is 8k which should be a safe default buffer size to use. It can be changed if needed.

I'd suggest increasing the default to 16KB, which may help decreasing CPU usage. The performance seems to be the same between 8 and 16KB buffer, but I'm wondering about the CPU usage.

With a buffer size of 8K and a data rate of 382MB/sec, the downloader will be doing up to 100K system calls per second (382MB*1024 / 8KB *2, one syscall to read and maybe another to write, possibly less than one per write because of write buffering in the JVM). That seems like a very high number, and given that syscalls have a moderate cost, this could result in unnecessary CPU usage.

If you don't see any drawback to increasing the buffer to 16KB, then this could reduce CPU usage, which is always good in a cloud environment. Even with 1000 download threads, 16KB buffers would be 16MB of RAM, not much.